sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx

```diff
@@ -0,0 +1,330 @@
+    tracks the latest execution state ready to be reported. Snapshot versions
+    are associated with values in the execution_journal and done_journal.
+    This value written by ExecutionJournalWriter, and read by the
+    ExecutionJournalReader.
+  """
+  def __init__(self):
+    self.latest_snapshot = 0
+    self.execution_journal = Journal()
+    self.done_journal = Journal()
+
+cdef class ExecutionJournalReader(object):
+  """Accounts sampled time to processed elements based on execution journal
+  entries.
+
+  Attributes:
+    shared_state: Shared journals with ExecutionJournalWriter.
+    execution_duration: Accumulated execution time per element. Once an element
+    has finished processing and execution time has been attributed, the total
+    execution time is reported via the counter and removed from the collection.
+
+  Args:
+    shared_state: Initialized in DataflowElementExecutionTracker.
+  """
+  def __init__(self, ReaderWriterState shared_state):
+    self.shared_state = shared_state
+    self.execution_duration = unordered_map[ElementExecutionPtr, int64_t]()
+
+  cdef void take_sample(self, int64_t sample_time,
+      unordered_map[CharPtr, vector[int64_t]]& counter_cache) nogil:
+    """Account the specified processing time duration to elements which have 
+    processed since the last sampling round, and update counter_cache for 
+    completed elements.
+    Args:
+      sample_time: Total sampling time since last time.
+      counter_cache: Initialized in DataflowElementExecutionTracker.
+    """
+    latest_snapshot = self.shared_state.latest_snapshot
+    self.attribute_processing_time(sample_time, latest_snapshot)
+    self.update_counter_cache(latest_snapshot, counter_cache)
+
+  cdef void attribute_processing_time(self, int64_t sample_time,
+                                      int64_t latest_snapshot) nogil:
+    """Attribute processing time to elements from execution_journal up to the 
+    specified snapshot. 
+    """
+    cdef int64_t total_execution = 0
+    cdef unordered_map[ElementExecutionPtr, int64_t] executions_per_element
+    cdef SnapshottedExecutionPtr snapshotted_ptr = NULL
+    # Calculate total execution counts and prune execution_journal.
+    while self.shared_state.execution_journal.journal.empty() is False:
+      if (self.shared_state.execution_journal.journal.front().snapshot
+          <= latest_snapshot):
+        total_execution += 1
+        snapshotted_ptr = self.shared_state.execution_journal.journal.front()
+        self.shared_state.execution_journal.journal.pop_front()
+        if snapshotted_ptr.execution_ptr == NULL:
+          continue
+        executions_per_element[snapshotted_ptr.execution_ptr] += 1
+      else:
+        break
+    # Keep the currently executing element in the journal.
+    # So its remaining execution time is counted in the next sampling round.
+    if snapshotted_ptr != NULL and snapshotted_ptr.snapshot == latest_snapshot:
```

Comment:
<!--thread_id:cc_182316149_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> > may cause some execution get less attribution time.
> 
> Is this temporary? And would it fix itself in the next round of counter updates?

<!--section:body-->
Unfortunately no. ```<``` will cause some execution get less attributing time in final result.